src/sage/schemes/riemann_surfaces/riemann_surface.py

```diff
@@ -3990,6 +4011,720 @@ def vectomat(v):
+    - ``T`` -- the transformation matrix such that S * N * T = M (if ``transformation`` is True)
+
+    INPUT:
+
+    - ``M`` -- a matrix with integer entries
+
+    EXAMPLES::
+
+        sage: from sage.schemes.riemann_surfaces.riemann_surface import poincare_form
+        sage: M = Matrix(ZZ, [
+        ....:     [1, 0, 0, 0, 0, -1, 0, -1],
+        ....:     [0, 1, 0, 0, -1, -1, -1, 0],
+        ....:     [0, 0, 1, 0, 0, 1, 0, 0],
+        ....:     [0, 0, 0, 1, 1, 0, 0, 0]])
+        sage: N, S, T = poincare_form(M, transformation=True)
+        sage: N
+        [1 0 0 0 0 0 0 0]
+        [0 1 0 0 0 0 0 0]
+        [0 0 1 0 2 0 0 0]
+        [0 0 0 1 0 2 0 0]
+        sage: bool(S * N * T == M)
+        True
+    """
+
+    M = M.change_ring(ZZ)
+    n_r, n_c = M.nrows(), M.ncols()
+    m = n_r // 2
+    g = n_c // 2
+
+    if M.nrows() % 2 != 0 or M.ncols() % 2 != 0:
+        raise ValueError("Matrix dimensions must be even")
+
+    if M.rank() != M.nrows():
+        raise ValueError("Matrix is not full rank")
+
+    I_g = Matrix.identity(ZZ, g)
+    zero_g = Matrix.zero(ZZ, g, g)
+    J = Matrix.block([[zero_g, I_g], [-I_g, zero_g]])
+
+    if (M * J * M.transpose()).det() == 0:
+        raise ValueError("MJM^T is singular")
+
+    def symplectic_matrix(type, A, g):
+        I = Matrix.identity(ZZ, g)
+        Z = Matrix.zero(ZZ, g, g)
+        if type == "1":
+            return block_matrix([[I, A], [Z, I]])
+        elif type == "2":
+            return block_matrix([[I, Z], [A, I]])
+        elif type == "3":
+            return block_matrix([[A, Z], [Z, A.transpose().inverse()]])
+        else:
+            raise ValueError(f"Invalid symplectic type: {type}")
+
+    def partner(col):
+        return col + g if col < g else col - g
+
+    S = Matrix.identity(ZZ, n_r)
+    T = Matrix.identity(ZZ, n_c)
+
+    # Step 1: Fix/check subdeterminants
+    gcd_subdets = GCD(
+        [
+            block_matrix([M[:, ci] for ci in com], ncols=2 * m).det()
+            for com in combinations(range(n_c), n_r)
+        ]
+    )
+
+    if gcd_subdets > 1:
+        D, U, V = M.smith_form()
+        D_sq = D[:n_r, :n_r]
+        L_full = U.inverse() * D_sq * U
+        H, Q = (L_full.transpose()).hermite_form(transformation=True)
+        L_triag = H.transpose()
+        M_curr = L_triag.inverse() * M
+        S = L_triag * S
+    else:
+        M_curr = M
+
+    # Step 2: Frobenius alternating form
+    MJMT = M_curr * J * M_curr.transpose()
+    delta = MJMT[0:m, m:2*m]
+    in_form = True
+    for i in range(m):
+        for j in range(m):
+            if (delta[i, j] != 0 and i != j) or (delta[i, j] == 0 and i == j) or (delta[i, j] < 0):
+                in_form = False
+                break
+        else:
+            continue
+        break
+    else:
+        delta_diag = [delta[i, i] for i in range(m)]
+        
+        for i in range(m - 1):
+            if delta_diag[i + 1] != 0 and delta_diag[i] % delta_diag[i + 1] != 0:
```

Comment:
At this point you are assuming you have the matrix in the form `[[0, D], [-D, 0]]` for diagonal `D`, and then you are checking the divisibility condition, but given that you know `M` is full rank you necessarily know here that `D` cannot have any zero entries, so this first logical check is redundant.